### PR TITLE
feat(asset): add asset variable query with query builder

### PR DESCRIPTION
**Pull request by kkerezsi**
# Pull Request

## 🤨 Rationale

In order to list assets as variables, a Variable Query Editor needs to be implemented. As per the requirements, the variable needs to support advanced filtering. 

image

## 👩‍💻 Implementation

- Added asset variable query editor
- Added variable query builder

## 🧪 Testing

- Manual testing
- Unit testing

## ✅ Checklist

- [x] This PR has a title that follows the [commit message format](https://github.com/ni/systemlink-grafana-plugins#commit-message-format).

### CODE REVIEW

File: src/datasources/asset/AssetDataSource.ts [hunk 1]
```diff
@@ -82,4 +84,11 @@ export class AssetDataSource extends DataSourceBase<AssetQuery, AssetDataSourceO
   getListAssetsSource(): ListAssetsDataSource {
     return this.listAssetsDataSource;
   }
+
+  async metricFindQuery(query: AssetQuery): Promise<MetricFindValue[]> {
+    console.log(query);
+    const assetFilter = query as ListAssetsQuery;
```
Comment (saradei-ni) [inline src/datasources/asset/AssetDataSource.ts:90]: `const assetFilter = assetFilter?.filter ?? '';`

this should probably be enough as well as we use only that property.

File: src/datasources/asset/AssetDataSource.ts [hunk 2]
```diff
@@ -82,4 +84,11 @@ export class AssetDataSource extends DataSourceBase<AssetQuery, AssetDataSourceO
...
+    const assetsResponse: AssetModel[] = await this.getListAssetsSource().queryAssets(assetFilter?.filter ?? '', 1000);
```
Comment (saradei-ni) [inline src/datasources/asset/AssetDataSource.ts:91]: `this.getListAssetsSource().queryAssets(assetFilter, QUERY_LIMIT);`

We could define a constant for 1k as it's the upper limit we allow.

**Review by saradei-ni — APPROVED** [review]

File: src/datasources/asset/components/variable-editor/AssetVariableQueryEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+type Props = QueryEditorProps<AssetDataSource, AssetQuery, AssetDataSourceOptions>;
+
+export function AssetVariableQueryEditor({ datasource, query, onChange }: Props) {
+  const [errorMsg] = useState<string | undefined>('');
+  const [workspaces, setWorkspaces] = useState<Workspace[]>([]);
+  const [systems, setSystems] = useState<SystemMetadata[]>([]);
+  const [areDependenciesLoaded, setAreDependenciesLoaded] = useState<boolean>(false);
+  const listAssetQuery: ListAssetsQuery = query as ListAssetsQuery;
+  const assetListDatasource = datasource.getListAssetsSource();
```
[the remark this answers is not in the document]
Comment (Catalin-web) [reply src/datasources/asset/components/variable-editor/AssetVariableQueryEditor.tsx]: After some more investigation I don't think this is possible

File: src/datasources/asset/AssetDataSource.ts [hunk 4]
```diff
@@ -82,4 +85,10 @@ export class AssetDataSource extends DataSourceBase<AssetQuery, AssetDataSourceO
   getListAssetsSource(): ListAssetsDataSource {
     return this.listAssetsDataSource;
   }
+
+  async metricFindQuery(query: AssetQuery): Promise<MetricFindValue[]> {
+    const assetFilter = (query as ListAssetsQuery)?.filter ?? '';
+    const assetsResponse: AssetModel[] = await this.listAssetsDataSource.queryAssets(assetFilter, QUERY_LIMIT);
+    return assetsResponse.map((asset: AssetModel) => ({ text: asset.name, value: `Asset.${asset.vendorName}.${asset.modelName}.${asset.serialNumber}` }));
```
Comment (Catalin-web) [inline src/datasources/asset/AssetDataSource.ts:92]: The value should not be the id?

Comment (Catalin-web) [reply src/datasources/asset/AssetDataSource.ts:92]: What happens if 2 assets have the same name?

**Review by Catalin-web — APPROVED** [review]

File: src/datasources/asset/AssetDataSource.ts (hunk 4 above)
Comment (kkerezsi) [reply src/datasources/asset/AssetDataSource.ts:92]: This was brought up already and was a request from @joshuaprewitt  to have it  in this form

**Comment by mure** [comment]
@kkerezsi Does this replace #71?

**Comment by kkerezsi** [comment]
@mure yes

File: src/datasources/asset/AssetDataSource.ts [hunk 5]
```diff
@@ -82,4 +87,15 @@ export class AssetDataSource extends DataSourceBase<AssetQuery, AssetDataSourceO
   getListAssetsSource(): ListAssetsDataSource {
     return this.listAssetsDataSource;
   }
+
+  async metricFindQuery(query: AssetQuery, options: LegacyMetricFindQueryOptions): Promise<MetricFindValue[]> {
```
Comment (mure) [inline src/datasources/asset/AssetDataSource.ts:91]: I would recommend defining a separate interface for the variable query like we do here: https://github.com/ni/systemlink-grafana-plugins/blob/a315a292ff6576996bbf89ea2a9fc1c912678870/src/datasources/system/SystemDataSource.ts#L84

**Review by mure — APPROVED** [review]

**Review by mure — APPROVED** [review]

**Pull request merged**